modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessExecuteAction.java

```diff
@@ -551,9 +551,11 @@ private static Response prepareRamIndex(Request request,
                     ParsedDocument parsedDocument = indexService.mapperService().documentMapper().parse(sourceToParse);
                     indexWriter.addDocuments(parsedDocument.docs());
                     try (IndexReader indexReader = DirectoryReader.open(indexWriter)) {
+                        final IndexSearcher searcher = new IndexSearcher(indexReader);
+                        searcher.setQueryCache(null);
                         final long absoluteStartMillis = System.currentTimeMillis();
                         QueryShardContext context =
-                            indexService.newQueryShardContext(0, indexReader, () -> absoluteStartMillis, null);
+                            indexService.newQueryShardContext(0, searcher, () -> absoluteStartMillis, null);
```

Comment:
out of curiosity: was this change necessary?

Reply:
For simplicity the `QueryShardContext` should expose an `IndexSearcher` that is configured with the correct similarity, cache, ... I can make this change in a different pr if you like ?

Reply:
it would be nice, unless that cause too much extra work on your end.

Reply:
I opened https://github.com/elastic/elasticsearch/pull/46584 to split this pr a bit.

Reply:
I merged #46584 and incorporated the changes in this pr, can you take another look ?